docs/reference/builder.md

```diff
@@ -954,6 +954,41 @@ For example:
 The output of the final `pwd` command in this `Dockerfile` would be
 `/path/$DIRNAME`
 
+## ARG
```

Comment:
@mapuri Cool PR and really popular. I think this section could be clarified a bit. Also, reading the discussion, the code, and other docs...expansion is warranted.  For example, there seems to be a relationship between `ENV` and `ARG` --- for example implied but not clear can i refer to an ARG in ENV. Implied but not clear, can I override an `ARG` by specifying an `ENV` of the same name.  It might be in the discussion but I didn't see it.  So, I touch on ENV here please address any technical inaccuracies I introduce.

Also, not clear on the predefined ARG values can I specify a default in my Dockerfile? Seems like I should be able to.  Finally, examples using the pipe just are hard to read for the terminally lazy reader of tech docs --- me and somehow I think I'm not alone. :grinning: 

---> https://gist.github.com/moxiegirl/d845ac58868cb2ec1036

## ARG


    ARG [--default=<value>] [--description=<value>] <arg name>

The `ARG` instruction defines a variable that users can pass at build-time to
the builder with the `docker run` command using the `--build-arg
<varname>=<value>` flag. If a user specifies a `--build-arg` but it is not
used by the builder, the build outputs a warning.

```
One or more build-args %v were not consumed, failing build.
```

The Dockerfile developer can define a single variable by specifying `ARG` once or many
variables by specifying `ARG` more than once. For example, a valid Dockerfile:

```
FROM busybox
ARG user
ARG buildno
...
```

A Dockerfile developer may optionally specify a default value and description
for an `ARG` instruction:

```
FROM busybox
ARG user
ARG buildno --default="XXX"
...
```

If an `ARG` value has a default and if there is no value passed at build-time, the
builder uses the default. The default value and the description are persisted in
the image metadata. Persistence makes these optional flags useful as
documentation. If an `ARG` does not contain either of these flags, the `ARG` is not
persisted.

You can pass variables in an `ARG` or an `ENV` instruction to specify variables
that are available to the `RUN` instruction. Users can only override the `ARG`
instruction from the `docker run` command line. If a you define a variable of
the same name in both an `ARG` and an `ENV` instruction, the `ENV` instruction
is used in the `RUN` context. And, unlike an `ARG` instructions, variables defined
in  `ENV` instructions always persist in the image metadata.

An `ARG` variable definition comes into effect from the line on which it is
defined in the `Dockerfile` not from the argument's use on the command-line or
elsewhere.  For example, consider this Dockerfile:

```
1 FROM busybox
2 USER ${user:-foo}
3 ARG user
4 USER $user
...
```
A user builds this file by calling:

```
$ docker build --build-arg user=foo1 Dockerfile
```

The `USER` at line 2 evaluates to `foo` as the `ARG user` is defined on the
subsquent line 3 line. The `USER` at line 4 evaluates to `foo1` as `user` is
defined and `foo1` was passed on the command line. Prior to its definition by an
`ARG` instruction, any use of a variable results in an empty string. 

Docker has set of predefined `ARG` variables that you can use. 

* `HTTP_PROXY`
* `HTTPS_PROXY`
* `http_proxy`
* `https_proxy`
* `FTP_PROXY`
* `NO_PROXY`

To use these, simply pass them on the command line using the `--build-arg
<varname>=<value>` flag. You can also define defaults for these in a Dockerfile.